examples/03_press_drag_release_tool/ex_press_drag_release_tool/main.rb

```diff
@@ -0,0 +1,182 @@
+        view.invalidate
+      end
+
+      def onLButtonDown(flags, x, y, view)
+        # Track where in screen space mouse is pressed down.
+        @mouse_down = Geom::Point3d.new(x, y)
+
+        if picked_first_point? && create_edge > 0
+          # When the user have picked a start point and then picks another point
+          # we create an edge and try to create new faces from that edge.
+          # Like the native tool we reset the tool if it created new faces.
+          reset_tool
+        else
+          # If no face was created we let the user chain new edges to the last
+          # input point.
+          @picked_first_ip.copy!(@mouse_ip)
+        end
+
+        update_ui
+        view.invalidate
+      end
+
+      def onLButtonUp(flags, x, y, view)
+        if @mouse_down.distance([x, y]) > DRAG_THRESHOLD
+          # Mouse is released far enough away from where it was pressed to be
+          # considered to be dragged there.
+
+          create_edge
+
+          # When drag-drawing, it feels odd to chain line drawing and we
+          # consistently reset the tool instead.
+          # You can try only calling this method if create_edge returns a
+          # positive number, to experience the other behavior.
+          reset_tool
+        end
+      end
+
+      # Here we have hard coded a special ID for the pencil cursor in SketchUp.
+      # Normally you would use `UI.create_cursor(cursor_path, 0, 0)` instead
+      # with your own custom cursor bitmap:
+      #
+      #   CURSOR_PENCIL = UI.create_cursor(cursor_path, 0, 0)
+      CURSOR_PENCIL = 632
+
+      def onSetCursor
+        # Note that `onSetCursor` is called frequently so you should not do much
+        # work here. At most you switch between different cursor representing
+        # the state of the tool.
+        UI.set_cursor(CURSOR_PENCIL)
+      end
+
+      def draw(view)
+        draw_preview(view)
+        @mouse_ip.draw(view) if @mouse_ip.display?
+      end
+
+      def getExtents
+        bounds = Geom::BoundingBox.new
+        bounds.add(picked_points)
+        bounds
+      end
+
+      private
+
+      def update_ui
+        if picked_first_point?
+          Sketchup.status_text = 'Select end point.'
+        else
+          Sketchup.status_text = 'Select start point.'
+        end
+
+        Sketchup.vcb_label = 'Length'
+      end
+
+      def reset_tool
+        @picked_first_ip.clear
+        update_ui
+      end
+
+      def picked_first_point?
+        @picked_first_ip.valid?
+      end
+
+      def picked_points
+        points = []
+        points << @picked_first_ip.position if picked_first_point?
+        points << @mouse_ip.position if @mouse_ip.valid?
+        points
+      end
+
+      def draw_preview(view)
+        points = picked_points
+        return unless points.size == 2
+        view.set_color_from_line(*points)
+        view.line_width = 1
+        view.line_stipple = ''
```

Comment:
Side note: We should add constants for this. Ditto for the point styles for `view.draw_points`. I keep copying over the same constants for every extension project.